server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/delivery/MailDispatcherTest.java

```diff
@@ -0,0 +1,259 @@
+            .state(state)
+            .build();
+        testee.dispatch(mail);
+
+        assertThat(mail.getState()).isEqualTo(state);
+    }
+
+    @Test
+    public void errorsShouldBeWellHandled() throws Exception {
+        MailDispatcher testee = MailDispatcher.builder()
+            .log(mock(Log.class))
+            .mailetContext(fakeMailContext)
+            .mailStorer(mailStorer)
+            .consume(true)
+            .build();
+        doThrow(new MessagingException())
+            .when(mailStorer)
+            .storeMail(any(MailAddress.class), any(MailAddress.class), any(Mail.class));
+
+        MimeMessage mimeMessage = new MimeMessage(Session.getDefaultInstance(new Properties()));
+        Multipart multipart = new MimeMultipart();
+        MimeBodyPart bodyPart = new MimeBodyPart();
+        bodyPart.setDataHandler(
+            new DataHandler(
+                new ByteArrayDataSource(
+                    "toto",
+                    "text/plain; charset=UTF-8")
+            ));
+        multipart.addBodyPart(bodyPart);
+        mimeMessage.setContent(multipart);
+        mimeMessage.saveChanges();
+
+        FakeMail mail = FakeMail.builder()
+            .sender(MailAddressFixture.OTHER_AT_JAMES)
+            .mimeMessage(mimeMessage)
+            .recipients(MailAddressFixture.ANY_AT_JAMES)
+            .state("state")
+            .build();
+        testee.dispatch(mail);
+
+        List<FakeMailContext.SentMail> actual = fakeMailContext.getSentMails();
+        FakeMailContext.SentMail expected = new FakeMailContext.SentMail.Builder()
+            .sender(MailAddressFixture.OTHER_AT_JAMES)
+            .recipients(ImmutableList.of(MailAddressFixture.ANY_AT_JAMES))
+            .state(Mail.ERROR).build();
+        assertThat(actual).containsOnly(expected);
+        assertThat(IOUtils.toString(actual.get(0).getMsg().getInputStream(), Charsets.UTF_8))
+            .contains("toto");
+    }
+
+    @Test
+    public void dispatchShouldUpdateReturnPath() throws Exception {
+        MailDispatcher testee = MailDispatcher.builder()
+            .log(mock(Log.class))
+            .mailetContext(fakeMailContext)
+            .mailStorer(mailStorer)
+            .consume(false)
+            .build();
+
+        FakeMail mail = FakeMail.builder()
+            .sender(MailAddressFixture.OTHER_AT_JAMES)
+            .recipients(MailAddressFixture.ANY_AT_JAMES)
+            .mimeMessage(new MimeMessage(Session.getDefaultInstance(new Properties())))
+            .state("state")
+            .build();
+        testee.dispatch(mail);
+
+        ArgumentCaptor<Mail> mailCaptor = ArgumentCaptor.forClass(Mail.class);
+        verify(mailStorer).storeMail(any(MailAddress.class), any(MailAddress.class), mailCaptor.capture());
+
+        assertThat(mailCaptor.getValue().getMessage().getHeader(RFC2822Headers.RETURN_PATH))
+            .containsExactly("<" + MailAddressFixture.OTHER_AT_JAMES +">");
```

Comment:
We have utility function somewhere, should we do that or just keep this one for clearly?